rustls/src/server/mod.rs

```diff
@@ -95,26 +96,47 @@ pub trait ProducesTickets : Send + Sync {
 /// How to choose a certificate chain and signing key for use
 /// in server authentication.
 pub trait ResolvesServerCert : Send + Sync {
-    /// Choose a certificate chain and matching key given any server DNS
-    /// name provided via SNI, any ALPN protocol names, and signature schemes
-    /// advertised by the client.
-    ///
-    /// `server_name` is `None` if the client did not supply an SNI name.
-    ///
-    /// `sigschemes` is replaced with a standard-specified default if the client
-    /// omitted this extension.
-    ///
-    /// `alpn_protocols` is `None` if the client did not include an ALPN extension.
-    ///
-    /// Return a `CertifiedKey` to continue the handshake.  A `CertifiedKey`
-    /// binds together a signing key and a certificate chain.
-    ///
+    /// Choose a certificate chain and matching key given simplified
+    /// ClientHello information.
+    /// 
     /// Return `None` to abort the handshake.
-    fn resolve(&self,
-               server_name: Option<webpki::DNSNameRef>,
-               sigschemes: &[SignatureScheme],
-               alpn_protocols: Option<&[&[u8]]>)
-               -> Option<sign::CertifiedKey>;
+    fn resolve(&self, client_hello: ClientHello) -> Option<sign::CertifiedKey>;
+}
+
+/// A struct representing the received Client Hello
+pub struct ClientHello<'a> {
+    server_name: Option<webpki::DNSNameRef<'a>>,
+    sigschemes: &'a [SignatureScheme],
+    alpn: Option<&'a[&'a[u8]]>,
+}
+
+impl<'a> ClientHello<'a> {
+    /// Creates a new ClientHello
+    pub fn new(server_name: Option<webpki::DNSNameRef<'a>>, sigschemes:  &'a [SignatureScheme],
+    alpn: Option<&'a[&'a[u8]]>)->Self {
+        ClientHello {server_name, sigschemes, alpn}
+    }
```

Comment:
If the goal is to ensure forwards-compatibility, perhaps this shouldn't be public?

Reply:
Good idea.